README.md
# SEO Writing Assistant SEMrush Custom Fields

The [SEMrush SEO Writing Assistant](https://wordpress.org/plugins/semrush-seo-writing-assistant/) plugin read only from post `title` and post `content` elements for the real time check.

This lightweight plugin allows you to use an arbitrary value for that check in addition to those values.

So, if you are using some kind of a page builders, field managers and so on, you have only to define a constant in your 
_wp_config.php_ file and eventually to write a little bit of JavaScript code in order to specify which fields.

Note that the plugin don't replace _SEMrush SEO Writing Assistant_ but is intended to be used together.

# Installation  

This section describes how to install the plugin and get it working.

1. Upload the plugin files to the `/wp-content/plugins/seo-writing-assistant-semrush-custom-fields` directory, or install the plugin through the WordPress _Plugins_ screen directly.
1. Activate the plugin through the _Plugins_ screen in WordPress.

# Usage

Once the plugin is activated you can configure it by defining following constants in your _wp-config.php_:

* `SWA_SEMRUSH_CUSTOM_FIELDS_PLUGIN_AUTOENABLE`

Automatically enable use of additional text value for SEMrush check. The default value is `false`.
This is the only configuration you have to do in order to have the plugin feature working.
For example:

```php
define( 'SWA_SEMRUSH_CUSTOM_FIELDS_PLUGIN_AUTOENABLE', true );
```
* `SWA_SEMRUSH_CUSTOM_FIELDS_PLUGIN_INTERVAL`

Allows to change the interval between automatic updates of the text used by SEMrush checks.
The default value is `5` seconds.For example:

```php
define( 'SWA_SEMRUSH_CUSTOM_FIELDS_PLUGIN_INTERVAL', 10 );
```
You can alernatavely control the plugins programmatically, by using 
`semrush_seo_writing_assistant_post_types` (optional) filter and `SeoWritingAssistantSEMrushCustomFields` JavaScript object (see below).

# API

## WordPress Hooks

### `swa_semrush_custom_fields_settings`

Filters plugin settings values.

```php
apply_filters( 'swa_semrush_custom_fields_settings', array $settings )
```

### `semrush_seo_writing_assistant_post_types`

Filters post types where to enable plugin features. Default to `array( 'post', 'page', 'product' )`.

```php
apply_filters( 'semrush_seo_writing_assistant_post_types', array $post_types )
```

The filter has the same name as the one used by _SEMrush SEO Writing Assistant_ plugin (since version 1.0.4) used for the same purpose.

## JavaScript Event

### `seo-writing-assistant-semrush`

This event allows you to update the text for SEMrush analysis.
The following example will update text for SEMrush check every 5 seconds, using `excerpt` and `my_custom_fields` custom field values (working both with block and classic editors):

```javascript
jQuery( document ).ready( function() {
	var swa = new SeoWritingAssistantSEMrushCustomFields( { interval: 5 } );
	jQuery( document ).on( 'seo-writing-assistant-semrush', function( event, data ) {
		data.html += jQuery( '#excerpt' ).val() + "\n"
			+ jQuery( '.editor-post-excerpt__textarea textarea').val() + "\n"
			+ jQuery( '#my_custom_fields' ).val();
	} )
} );
```
Note that if you have previously defined constant `SWA_SEMRUSH_CUSTOM_FIELDS_PLUGIN_AUTOENABLE` to `true`, 
you don't have to create new `SeoWritingAssistantSEMrushCustomFields` object.

## JavaScript API

### `setHtml()`

The following example will update programmatically the text used by SEMrush check with an arbitrary value:

```javascript
jQuery( document ).ready( function() {
	var swa = new SeoWritingAssistantSEMrushCustomFields( { interval: 0 } );
	swa.setHtml( 'foo baz bar' );
} );
```
## Advanced Custom Fields

The plugin checks if _Advanced Custom Fields_ is active in your WordPress instance and adds a custom setting to each field 
(only for `input` within _text_, _radio_ and _checkbox_ type, `textarea` and `select` elements) allowing you to add his value 
to SEMrush check without having to write any additional JavaScript code.

## Autodiscovery

The plugin also automatically add to text used by SEMrush checks, the values of all HTML elements
with `swa-scf` class or `data-swa-scf` attribute.
This will applied only for `input` (within _text_, _radio_ and _checkbox_ type), `textarea`, and `select` elements.

# Screenshots 

## Advanced Custom Fields settings

The Advanced Custom Fields custom setting.

![Advanced Custom Fields custom setting](https://raw.githubusercontent.com/enrico-sorcinelli/seo-writing-assistant-semrush-custom-fields/master/assets-wp/screenshot-1.png)

# Frequently Asked Questions

## Does it work with Gutenberg / Block editor?

Yes.

# License: GPLv2

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [1.1.0] - 2019-06-16
### Added
 - Support for Advanced Custom Fields.
 - Autodiscovery feature.
 - Plugin settings and `swa_semrush_custom_fields_settings` filter.
 - Italian translations.

### Changed
 - Updated documentation.

## [1.0.3] - 2019-02-14
### Changed
 - Refactored accordying to the new plugin slug.

## [1.0.2] - 2019-02-12
### Fixed
 - PHPCS for VIP.

## [1.0.1] - 2019-02-09
### Added
 - Check for _SEMrush SEO Writing Assistant_ plugin status.

## [1.0.0] - 2019-02-08
### Added
 - First public release


tests/phpunit/bootstrap.php
<?php
/**
 * PHPUnit bootstrap file
 *
 * @package semrush-swa-custom-fields
 */

$_tests_dir = getenv( 'WP_TESTS_DIR' );

if ( ! $_tests_dir ) {
	$_tests_dir = rtrim( sys_get_temp_dir(), '/\\' ) . '/wordpress-tests-lib';
}

// Allow to override PHP executable used by WordPress Test framework.
$_php_binary = getenv( 'WP_PHP_BINARY' );
if ( ! empty( $_php_binary ) ) {
	define( 'WP_PHP_BINARY', $_php_binary );
}

if ( ! file_exists( $_tests_dir . '/includes/functions.php' ) ) {
	echo "Could not find $_tests_dir/includes/functions.php, have you run bin/install-wp-tests.sh ?" . PHP_EOL;
	exit( 1 );
}

// Give access to tests_add_filter() function.
require_once $_tests_dir . '/includes/functions.php';

/**
 * Manually load the plugin being tested.
 */
function _manually_load_plugin() {
	require dirname( dirname( __FILE__ ) ) . '../../seo-writing-assistant-semrush-custom-fields.php';
}
tests_add_filter( 'muplugins_loaded', '_manually_load_plugin' );

// Start up the WP testing environment.
require $_tests_dir . '/includes/bootstrap.php';


tests/phpunit/tests/test-sample.php
<?php
/**
 * Class SampleTest
 *
 * @package semrush-swa-custom-fields
 */

/**
 * Sample test case.
 */
class SampleTest extends WP_UnitTestCase {

	/**
	 * A single example test.
	 */
	function test_sample() {
		// Replace this with some actual testing code.
		$this->assertTrue( true );
	}
}


assets/js/semrush-swa-custom-fields.js
/**
 * @file This file contains SEMrush Seo Writing Assistant Custom Fields JavaScript class
 * @author Enrico Sorcinelli 
 * @version 1.0.0
 * @title SEMrush Seo Writing Assistant Custom Fields
 */

/**
 * Allow to pass to SEMrush SWA arbirtrary html in addition to post content.
 *
 * @param {Object}  args - The arguments has following properties.
 * @param {boolean} [args.debug=false] - Debug mode. If `true`, it will print on `console.log` (if defined) some debug useful informations.
 * @param {number}  [args.interval=5] - Sets interval, in seconds, for event trigger. Set to `0` to turn off.
 *
 * @fires `semrush-seo-writing-assistant`
 *
 * @constructor
 */
SEMrushSeoWritingAssistantCustomFields = function( args ) {

	// Default arguments.
	args = $.extend(true, {
			debug: false,
			interval: 5
		},
		args
	);

	// Check for SEMrush plugin.
	if ( $( '#swa-meta-box' ).length < 1 ) {
		return;
	}

	// Create textarea element if necessary.
	this.$cmsms_content_composer_text = $( '#cmsms_content_composer_text' );
	if ( this.$cmsms_content_composer_text.length < 1 ) {
		this.$cmsms_content_composer_text = $( '<textarea id="cmsms_content_composer_text" style="display: none;"></textarea>' );
		$( 'body' ).append( this.$cmsms_content_composer_text );
	}

	// Create interval.
	if ( args.interval > 0 ) {
		this.interval = setInterval( this.trigger.bind( this ), args.interval * 1000 );
	}
};

/**
 * Class methods.
 */
SEMrushSeoWritingAssistantCustomFields.prototype = {

	/**
	 * Trigger `semrush-seo-writing-assistant` event.
	 */
	'trigger': function() {
		var data = {};
		$( window.document ).trigger( 'semrush-seo-writing-assistant', [ data ] );
		this.setHtml( data.html );
	},

	/**
	 * Allow to add custom HTML.
	 *
	 * @param {String} html - Arbitrary HTML added for analisys.
	 */
	'setHtml': function( html ) {
		this.$cmsms_content_composer_text.html( html );
	}
};


assets/js/swa-semrush-custom-fields.js
/**
 * @file This file contains Seo Writing Assistant SEMrush Custom Fields JavaScript class
 * @author Enrico Sorcinelli 
 * @version 1.1.0
 * @title SEO Writing Assistant SEMrush Custom Fields
 */

/**
 * Allow to pass to SEMrush SWA arbirtrary html in addition to post content.
 *
 * @param {Object}  args - The arguments has following properties.
 * @param {boolean} [args.debug=false] - Debug mode. If `true`, it will print on `console.log` (if defined) some debug useful informations.
 * @param {number}  [args.interval=5] - Sets interval, in seconds, for event trigger. Set to `0` to turn off.
 *
 * @fires `seo-writing-assistant-semrush`
 *
 * @constructor
 */
SeoWritingAssistantSEMrushCustomFields = function( args ) {

	// Default arguments.
	args = $.extend(true, {
			debug: false,
			interval: 5
		},
		args
	);

	this.params = {
		debug: args.debug,
		interval: args.interval
	};

	// Check for SEMrush plugin.
	if ( $( '#swa-meta-box' ).length < 1 ) {
		return;
	}

	// Create textarea element if necessary.
	this.$cmsms_content_composer_text = $( '#cmsms_content_composer_text' );
	if ( this.$cmsms_content_composer_text.length < 1 ) {
		this.$cmsms_content_composer_text = $( '<textarea id="cmsms_content_composer_text" style="display: none;"></textarea>' );
		$( 'body' ).append( this.$cmsms_content_composer_text );
	}

	// Create interval.
	if ( this.params.interval > 0 ) {
		this.interval = setInterval( this.trigger.bind( this ), this.params.interval * 1000 );
	}

	// Add listener for elements with `swa-scf` class or `data-swa-scf` attribute.
	$( document ).on( 'seo-writing-assistant-semrush', function( event, data ) {

		$( '.swa-scf, [data-swa-scf]' ).each( function( i, e ) {

			if ( 'select' === $( e ).prop( 'tagName' ).toLowerCase() ) {
				data.html += "\n" + $( e ).val();
				return;
			}
			switch ( $( e ).prop('type') ) {
				case 'text' :
				case 'textarea' :
					data.html += "\n" + $( e ).val();
					break;
				case 'radio' :
				case 'checkbox' :
					if ( $( e ).is( ':checked' ) ) {
						data.html += "\n" + $( e ).val();
					}
					break;
				default:
					break;
			}
		} );

	} );

};

/**
 * Class methods.
 */
SeoWritingAssistantSEMrushCustomFields.prototype = {

	/**
	 * Trigger `seo-writing-assistant-semrush` event.
	 */
	'trigger': function() {
		var data = { html: '' };
		$( window.document ).trigger( 'seo-writing-assistant-semrush', [ data ] );
		this.setHtml( data.html );
	},

	/**
	 * Allow to add custom HTML.
	 *
	 * @param {String} html - Arbitrary HTML added for analisys.
	 */
	'setHtml': function( html ) {
		this.$cmsms_content_composer_text.html( html );
	}
};

// Auto-enable object.
$( document ).ready( function() {
	if ( swa_semrush_custom_fields_i18n.enable ) {
		swa_semrush_custom_fields = new SeoWritingAssistantSEMrushCustomFields( { interval: swa_semrush_custom_fields_i18n.interval } );
	}
} );

semrush-seo-writing-assistant-custom-fields.php
<?php
/**
 * SEMrush SEO Writing Assistant Custom Fields plugin for WordPress
 *
 * @package semrush-swa-custom-fields
 *
 * Plugin Name: SEMrush SEO Writing Assistant Custom Fields
 * Plugin URI:  https://github.com/enrico-sorcinelli/semrush-seo-writing-assistant-custom-fields
 * Description: A WordPress plugin that allows SEMrush SEO Writing Assistant working with custom fields
 * Author:      Enrico Sorcinelli
 * Author URI:  https://github.com/enrico-sorcinelli/semrush-seo-writing-assistant-custom-fields/graphs/contributors
 * Text Domain: semrush-swa-custom-fields
 * Domain Path: /languages/
 * Version:     1.0.2
 * License:     GPLv2 or later
 * License URI: https://www.gnu.org/licenses/gpl-2.0.html
 */

// Check running WordPress instance.
if ( ! defined( 'ABSPATH' ) ) {
	header( 'HTTP/1.1 404 Not Found' );
	exit();
}

if ( ! class_exists( 'SEMrush_SEO_Writing_Assistant_Custom_Fields' ) ) {

	// Plugins constants.
	define( 'SEMRUSH_SWA_CUSTOM_FIELDS_PLUGIN_VERSION', '1.0.2' );
	define( 'SEMRUSH_SWA_CUSTOM_FIELDS_PLUGIN_BASEDIR', dirname( __FILE__ ) );
	define( 'SEMRUSH_SWA_CUSTOM_FIELDS_PLUGIN_BASEURL', plugin_dir_url( __FILE__ ) );

	// Enable debug prints on error_log (only when WP_DEBUG is true).
	if ( ! defined( 'SEMRUSH_SWA_CUSTOM_FIELDS_PLUGIN_DEBUG' ) ) {
		define( 'SEMRUSH_SWA_CUSTOM_FIELDS_PLUGIN_DEBUG', false );
	}

	require_once( SEMRUSH_SWA_CUSTOM_FIELDS_PLUGIN_BASEDIR . '/php/class-semrush-seo-writing-assistant-custom-fields.php' );

	/**
	 * Init the plugin.
	 *
	 * Define SEMRUSH_SWA_CUSTOM_FIELDS_PLUGIN_DEBUG to false in your <i>wp-config.php</i> to disable.
	 *
	 * @return void
	 */
	function semrush_swa_custom_fields_plugin_init() {

		if ( defined( 'SEMRUSH_SWA_CUSTOM_FIELDS_PLUGIN_AUTOENABLE' ) && SEMRUSH_SWA_CUSTOM_FIELDS_PLUGIN_AUTOENABLE === false ) {
			return;
		}

		// Instantiate our plugin class and add it to the set of globals.
		$GLOBALS['semrush_swa_custom_fields_plugin'] = SEMrush_SEO_Writing_Assistant_Custom_Fields::get_instance( array( 'debug' => SEMRUSH_SWA_CUSTOM_FIELDS_PLUGIN_DEBUG && WP_DEBUG ) );
	}

	// Activate the plugin once all plugin have been loaded.
	add_action( 'plugins_loaded', 'semrush_swa_custom_fields_plugin_init' );

	// Activation hook.
	register_activation_hook( __FILE__, array( 'SEMrush_SEO_Writing_Assistant_Custom_Fields', 'plugin_activation' ) );
}


seo-writing-assistant-semrush-custom-fields.php
<?php
/**
 * SEMrush SEO Writing Assistant Custom Fields plugin for WordPress
 *
 * @package swa-semrush-custom-fields
 *
 * Plugin Name: SEO Writing Assistant SEMrush Custom Fields
 * Plugin URI:  https://github.com/enrico-sorcinelli/seo-writing-assistant-semrush-custom-fields
 * Description: A WordPress plugin that allows SEMrush SEO Writing Assistant working with custom fields
 * Author:      Enrico Sorcinelli
 * Author URI:  https://github.com/enrico-sorcinelli/seo-writing-assistant-semrush-custom-fields/graphs/contributors
 * Text Domain: swa-semrush-custom-fields
 * Domain Path: /languages/
 * Version:     1.1.0
 * License:     GPLv2 or later
 * License URI: https://www.gnu.org/licenses/gpl-2.0.html
 */

// Check running WordPress instance.
if ( ! defined( 'ABSPATH' ) ) {
	header( 'HTTP/1.1 404 Not Found' );
	exit();
}

if ( ! class_exists( 'SEO_Writing_Assistant_SEMrush_Custom_Fields' ) ) {

	// Plugins constants.
	define( 'SWA_SEMRUSH_CUSTOM_FIELDS_PLUGIN_VERSION', '1.1.0' );
	define( 'SWA_SEMRUSH_CUSTOM_FIELDS_PLUGIN_BASEDIR', dirname( __FILE__ ) );
	define( 'SWA_SEMRUSH_CUSTOM_FIELDS_PLUGIN_BASEURL', plugin_dir_url( __FILE__ ) );

	// Enable debug prints on error_log (only when WP_DEBUG is true).
	if ( ! defined( 'SWA_SEMRUSH_CUSTOM_FIELDS_PLUGIN_DEBUG' ) ) {
		define( 'SWA_SEMRUSH_CUSTOM_FIELDS_PLUGIN_DEBUG', false );
	}

	require_once( SWA_SEMRUSH_CUSTOM_FIELDS_PLUGIN_BASEDIR . '/php/class-seo-writing-assistant-semrush-custom-fields.php' );

	/**
	 * Init the plugin.
	 */
	function swa_semrush_custom_fields_plugin_init() {

		// Instantiate our plugin class and add it to the set of globals.
		$GLOBALS['swa_semrush_custom_fields_plugin'] = SEO_Writing_Assistant_SEMrush_Custom_Fields::get_instance( array( 'debug' => SWA_SEMRUSH_CUSTOM_FIELDS_PLUGIN_DEBUG && WP_DEBUG ) );
	}

	// Activate the plugin once all plugin have been loaded.
	add_action( 'plugins_loaded', 'swa_semrush_custom_fields_plugin_init' );

	// Activation hook.
	register_activation_hook( __FILE__, array( 'SEO_Writing_Assistant_SEMrush_Custom_Fields', 'plugin_activation' ) );
}


php/class-semrush-seo-writing-assistant-custom-fields.php
<?php
/**
 * Plugin base class.
 *
 * @package semrush-swa-custom-fields
 * @author Enrico Sorcinelli
 */

// Check running WordPress instance.
if ( ! defined( 'ABSPATH' ) ) {
	header( 'HTTP/1.1 404 Not Found' );
	exit();
}

/**
 * Plugin base class.
 */
class SEMrush_SEO_Writing_Assistant_Custom_Fields {

	/**
	 * Instance settings.
	 *
	 * @var array
	 */
	private $settings = array();

	/**
	 * Prefix used for options and postmeta fields, DOM IDs and DB tables.
	 *
	 * @var string
	 */
	private static $prefix = 'semrush_swa_custom_fields_plugin_';

	/**
	 * Instance of this class.
	 *
	 * @var object
	 */
	public static $instance;

	/**
	 * Plugin class constructor.
	 *
	 * @param array $args {
	 *    Constructor arguments list.
	 *
	 *    @type boolean $debug Default value is `false`.
	 * }
	 *
	 * @return object
	 */
	public function __construct( $args = array() ) {

		$this->settings = wp_parse_args(
			array(
				'debug' => false,
			),
			$args
		);

		// Check and load needed compoments.
		if ( is_admin() ) {
			$this->require_components();
		}
	}

	/**
	 * Get the singleton instance of this class.
	 *
	 * @param array $args Constructor arguments list.
	 *
	 * @return object
	 */
	public static function get_instance( $args = array() ) {
		if ( ! ( self::$instance instanceof self ) ) {
			self::$instance = new self( $args );
		}
		return self::$instance;
	}

	/**
	 * This function will include core files before the theme's functions.php
	 * file has been excecuted.
	 */
	public function require_components() {

		// For plugin checks.
		if ( ! function_exists( 'is_plugin_active_for_network' ) ) {
			require_once( ABSPATH . 'wp-admin/includes/plugin.php' );
		}

		// Check for SEMrush SEO Writing Assistant plugin.
		if ( ! is_plugin_active( 'semrush-seo-writing-assistant/semrush-seo-writing-assistant.php' ) ) {
			return;
		}

		// Add JS to post pages.
		if ( ! self::is_ajax_request() ) {
			add_action( 'admin_print_scripts-post.php', array( $this, 'load_javascript' ), 10, 0 );
			add_action( 'admin_print_scripts-post-new.php', array( $this, 'load_javascript' ), 10, 0 );
		}
	}

	/**
	 * Load JavaScript files.
	 *
	 * @return void
	 */
	public function load_javascript() {

		global $post_type;

		/**
		 * Filter post types array where to add files.
		 *
		 * @param array $post_types. Default to `array( 'post', 'page', 'product' )`.
		 *
		 * @return array
		 */
		$allowed_post_types = apply_filters( 'semrush_seo_writing_assistant_post_types', array( 'post', 'page', 'product' ) );

		if ( empty( $post_type ) || ! in_array( $post_type, $allowed_post_types, true ) ) {
			return;
		}

		wp_enqueue_script(
			self::$prefix . 'js',
			SEMRUSH_SWA_CUSTOM_FIELDS_PLUGIN_BASEURL . '/assets/js/semrush-swa-custom-fields.js',
			array(),
			SEMRUSH_SWA_CUSTOM_FIELDS_PLUGIN_VERSION,
			false
		);
	}

	/**
	 * Check for AJAX request.
	 *
	 * @return boolean
	 */
	public static function is_ajax_request() {

		if (
			// PHPCS:ignore.
			( ! empty( $_SERVER['HTTP_X_REQUESTED_WITH'] ) && strtolower( $_SERVER['HTTP_X_REQUESTED_WITH'] ) === 'xmlhttprequest' )
			|| ( defined( 'DOING_AJAX' ) && DOING_AJAX )
		) {
			return true;
		}

		return false;
	}

	/**
	 * Plugin activation hook.
	 *
	 * @return void
	 */
	public static function plugin_activation() {
		if ( ! function_exists( 'is_plugin_active_for_network' ) ) {
			require_once( ABSPATH . 'wp-admin/includes/plugin.php' );
		}
		if ( current_user_can( 'activate_plugins' )
			&& ! is_plugin_active( 'semrush-seo-writing-assistant/semrush-seo-writing-assistant.php' ) ) {

			// Throw an error in the WordPress admin console.
			die(
				sprintf(
					'<p style="color: #444; font-size: 13px; line-height: 1.4em;font-family: -apple-system, BlinkMacSystemFont, \'Segoe UI\', Roboto, Oxygen-Sans, Ubuntu, Cantarell, \'Helvetica Neue\', sans-serif;">' . __( 'The <i>%$1s</i> plugin requires %$2s plugin to be active.', 'semrush-swa-custom-fields' ) . '</p>',
					'SEMrush SEO Writing Assistant Custom Fields',
					'<a target="_blank" href="' . esc_url( 'https://wordpress.org/plugins/semrush-seo-writing-assistant/' ) . '">SEMrush SEO Writing Assistant</a>'
				)
			); // WPCS: XSS ok.
		}
	}
}


php/class-seo-writing-assistant-semrush-custom-fields.php
<?php
/**
 * Plugin base class.
 *
 * @package swa-semrush-custom-fields
 * @author Enrico Sorcinelli
 */

// Check running WordPress instance.
if ( ! defined( 'ABSPATH' ) ) {
	header( 'HTTP/1.1 404 Not Found' );
	exit();
}

/**
 * Plugin base class.
 */
class SEO_Writing_Assistant_SEMrush_Custom_Fields {

	/**
	 * Instance settings.
	 *
	 * @var array
	 */
	private $settings = array();

	/**
	 * Plugin settings.
	 *
	 * @var array
	 */
	private $plugin_settings;

	/**
	 * Prefix used for options and postmeta fields, DOM IDs and DB tables.
	 *
	 * @var string
	 */
	private static $prefix = 'swa_semrush_custom_fields_';

	/**
	 * Instance of this class.
	 *
	 * @var object
	 */
	public static $instance;

	/**
	 * Plugin class constructor.
	 *
	 * @param array $args {
	 *    Constructor arguments list.
	 *
	 *    @type boolean $debug Default value is `false`.
	 * }
	 *
	 * @return object
	 */
	public function __construct( $args = array() ) {

		$this->settings = wp_parse_args(
			array(
				'debug' => false,
			),
			$args
		);

		// Load plugin text domain.
		load_plugin_textdomain( 'swa-semrush-custom-fields', false, dirname( plugin_basename( __FILE__ ) ) . '/../languages/' );

		// Check and load needed compoments.
		if ( is_admin() ) {
			$this->require_components();
		}
	}

	/**
	 * Get the singleton instance of this class.
	 *
	 * @param array $args Constructor arguments list.
	 *
	 * @return object
	 */
	public static function get_instance( $args = array() ) {
		if ( ! ( self::$instance instanceof self ) ) {
			self::$instance = new self( $args );
		}
		return self::$instance;
	}

	/**
	 * This function will include core files before the theme's functions.php
	 * file has been excecuted.
	 */
	public function require_components() {

		// For plugin checks.
		if ( ! function_exists( 'is_plugin_active_for_network' ) ) {
			require_once( ABSPATH . 'wp-admin/includes/plugin.php' );
		}

		// Check for SEMrush SEO Writing Assistant plugin.
		if ( ! is_plugin_active( 'semrush-seo-writing-assistant/semrush-seo-writing-assistant.php' ) ) {
			return;
		}

		// Get plugin settings.
		$this->plugin_settings = $this->get_plugin_default_settings();

		// Add JS to post pages.
		if ( ! self::is_ajax_request() ) {
			add_action( 'admin_print_scripts-post.php', array( $this, 'load_javascript' ), 10, 0 );
			add_action( 'admin_print_scripts-post-new.php', array( $this, 'load_javascript' ), 10, 0 );
		}

		// Check for ACF.
		if ( function_exists( 'get_field' ) ) {
			add_action( 'acf/render_field_settings/type=text', array( $this, 'acf_render_field_settings' ) );
			add_action( 'acf/render_field_settings/type=textarea', array( $this, 'acf_render_field_settings' ) );
			add_action( 'acf/render_field_settings/type=select', array( $this, 'acf_render_field_settings' ) );
			add_action( 'acf/render_field_settings/type=radio', array( $this, 'acf_render_field_settings' ) );
			add_action( 'acf/render_field_settings/type=checkbox', array( $this, 'acf_render_field_settings' ) );
			add_filter( 'acf/prepare_field', array( $this, 'acf_prepare_field' ) );
		}
	}

	/**
	 * Load JavaScript files.
	 *
	 * @return void
	 */
	public function load_javascript() {

		global $post_type;

		/**
		 * Filter post types array where to add files.
		 *
		 * @param array $post_types. Default to `array( 'post', 'page', 'product' )`.
		 *
		 * @return array
		 */
		$allowed_post_types = apply_filters( 'semrush_seo_writing_assistant_post_types', array( 'post', 'page', 'product' ) );

		if ( empty( $post_type ) || ! in_array( $post_type, $allowed_post_types, true ) ) {
			return;
		}

		// Add JavaScript.
		wp_enqueue_script(
			self::$prefix . 'js',
			SWA_SEMRUSH_CUSTOM_FIELDS_PLUGIN_BASEURL . '/assets/js/swa-semrush-custom-fields.js',
			array(),
			SWA_SEMRUSH_CUSTOM_FIELDS_PLUGIN_VERSION,
			false
		);

		// Localization.
		wp_localize_script(
			self::$prefix . 'js',
			self::$prefix . 'i18n',
			array(
				'enable'   => $this->plugin_settings['enable'],
				'interval' => $this->plugin_settings['interval'],
			)
		);
	}

	/**
	 * Add custom ACF settings.
	 *
	 * @param array $field Field data.
	 */
	function acf_render_field_settings( $field ) {

		acf_render_field_setting(
			$field,
			array(
				'label'        => __( 'SEMrush', 'swa-semrush-custom-fields' ),
				'instructions' => __( 'Add field value to the text used by SEMrush Writing Assistant analysis', 'swa-semrush-custom-fields' ),
				'name'         => 'swa_scf',
				'type'         => 'true_false',
				'ui'           => 1,
			)
		);
	}

	/**
	 * Add class to ACF fields.
	 *
	 * @param array $field Field data.
	 *
	 * @return mixed
	 */
	function acf_prepare_field( $field ) {
		if ( ! empty( $field['swa_scf'] ) ) {
			$field['class'] .= ' swa-scf';
		}
		return $field;
	}

	/**
	 * Check for AJAX request.
	 *
	 * @return boolean
	 */
	public static function is_ajax_request() {

		if (
			// PHPCS:ignore.
			( ! empty( $_SERVER['HTTP_X_REQUESTED_WITH'] ) && strtolower( $_SERVER['HTTP_X_REQUESTED_WITH'] ) === 'xmlhttprequest' )
			|| ( defined( 'DOING_AJAX' ) && DOING_AJAX )
		) {
			return true;
		}

		return false;
	}

	/**
	 * Get plugin options settings.
	 *
	 * @return array
	 */
	private function get_plugin_default_settings() {

		$settings = array(
			'enable'   => ( defined( 'SWA_SEMRUSH_CUSTOM_FIELDS_PLUGIN_AUTOENABLE' ) && SWA_SEMRUSH_CUSTOM_FIELDS_PLUGIN_AUTOENABLE === true ) ? true : false,
			'interval' => defined( 'SWA_SEMRUSH_CUSTOM_FIELDS_PLUGIN_INTERVAL' ) ? absint( SWA_SEMRUSH_CUSTOM_FIELDS_PLUGIN_INTERVAL ) : 5,
		);

		/**
		 * Filter plugin settings.
		 *
		 * @param array $settings Plugin settings.
		 *
		 * @return array
		 */
		$settings = apply_filters( self::$prefix . 'settings', $settings );

		return $settings;
	}

	/**
	 * Plugin activation hook.
	 *
	 * @return void
	 */
	public static function plugin_activation() {
		if ( ! function_exists( 'is_plugin_active_for_network' ) ) {
			require_once( ABSPATH . 'wp-admin/includes/plugin.php' );
		}
		if ( current_user_can( 'activate_plugins' )
			&& ! is_plugin_active( 'semrush-seo-writing-assistant/semrush-seo-writing-assistant.php' ) ) {

			// Load translations.
			load_plugin_textdomain( 'swa-semrush-custom-fields', false, dirname( plugin_basename( __FILE__ ) ) . '/../languages/' );

			// Throw an error in the WordPress admin console.
			die(
				sprintf(
					/* translators: 1 current plugin name, 2 plugin dependency name */'<p style="color: #444; font-size: 13px; line-height: 1.4em;font-family: -apple-system, BlinkMacSystemFont, \'Segoe UI\', Roboto, Oxygen-Sans, Ubuntu, Cantarell, \'Helvetica Neue\', sans-serif;">' . __( 'The %1$s plugin requires %2$s plugin to be active.', 'swa-semrush-custom-fields' ) . '</p>',
					'<i>SEO Writing Assistant SEMrush  Custom Fields</i>',
					'<a target="_blank" href="' . esc_url( 'https://wordpress.org/plugins/semrush-seo-writing-assistant/' ) . '">SEMrush SEO Writing Assistant</a>'
				)
			); // WPCS: XSS ok.
		}
	}
}
